### PR TITLE
CI: tests from go1.17

### PR DESCRIPTION
**Pull request by SuperQ**
## Summary
Test old Go versions all the way back to 1.17.

### CODE REVIEW

**Comment by SuperQ** [comment]
@dolmen Due to use of `reflect`, the minimum Go version is already 1.17.

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -17,3 +19,19 @@ jobs:
...
+      matrix:
+        go_version:
+          - "1.17"
+          - "1.18"
+          - "1.19"
+          - "1.20"
+    steps:
+      - uses: actions/checkout@v3
+      - name: Setup Go
+        uses: actions/setup-go@v3.2.0
```
Comment (dolmen) [inline .github/workflows/main.yml:34]: Must use v4.

File: README.md [hunk 2]
```diff
@@ -350,7 +350,7 @@ To update Testify to the latest version, use `go get -u github.com/stretchr/test
 Supported go versions
 ==================
 
-We currently support the most recent major Go versions from 1.19 onward.
+We currently support the most recent major Go versions from 1.17 onward.
```
Comment (dolmen) [inline README.md:353]: Do not change this right now. I want the CI to run first with the old versions.

Also, the versions supported is not a light policy change and other co-maintainers must be involved.

**Review by dolmen — CHANGES_REQUESTED** [review]

File: .github/workflows/main.yml (hunk 1 above)
Comment (SuperQ) [reply .github/workflows/main.yml]: Done

File: README.md (hunk 2 above)
Comment (SuperQ) [reply README.md]: Sure, as you can see, the CI is running for the old versions.

File: .github/workflows/main.yml [hunk 3]
```diff
@@ -6,14 +6,32 @@ jobs:
...
     strategy:
       matrix:
-        go_version: ["1.20", "1.19"]
+        go_version:
+          - "1.19"
+          - "1.20"
     steps:
       - uses: actions/checkout@v3
       - name: Setup Go
-        uses: actions/setup-go@v3.2.0
```
Comment (dolmen) [inline .github/workflows/main.yml:13]: Do not change here. That is #1408. I prefer that we keep those changes in separate commits. Especially as in CI results for #1408 we are seeing issues.

**Review by dolmen — CHANGES_REQUESTED** [review]

**Comment by dolmen** [comment]
@SuperQ 
> @dolmen Due to use of reflect, the minimum Go version is already 1.17.

What new features of `reflect` from go1.17 have you seen used here?

**Comment by SuperQ** [comment]
When I was doing my testing [this error happens in 1.16 and older](https://github.com/stretchr/testify/actions/runs/5464887276/jobs/9947538517).

```
Run go test -v -race ./...
go: downloading github.com/davecgh/go-spew v1.1.1
go: downloading github.com/pmezard/go-difflib v1.0.0
go: downloading github.com/stretchr/objx v0.5.0
go: downloading gopkg.in/yaml.v3 v3.0.1
# github.com/stretchr/testify/assert
Error: assert/assertions.go:9[4](https://github.com/stretchr/testify/actions/runs/5464887276/jobs/9947538517#step:4:5):23: field.IsExported undefined (type reflect.StructField has no field or method IsExported)
# github.com/stretchr/testify/assert [github.com/stretchr/testify/assert.test]
Error: assert/assertions.go:[9](https://github.com/stretchr/testify/actions/runs/5464887276/jobs/9947538517#step:4:10)4:23: field.IsExported undefined (type reflect.StructField has no field or method IsExported)
FAIL	github.com/stretchr/testify [build failed]
FAIL	github.com/stretchr/testify/assert [build failed]
FAIL	github.com/stretchr/testify/mock [build failed]
FAIL	github.com/stretchr/testify/require [build failed]
FAIL	github.com/stretchr/testify/suite [build failed]
Error: Process completed with exit code 2.
```

File: .github/workflows/main.yml (hunk 3 above)
Comment (SuperQ) [reply .github/workflows/main.yml:13]: Reverted

**Comment by dolmen** [comment]
@SuperQ Good to know. Thank you.

**Review by dolmen — DISMISSED** [review]

**Comment by SuperQ** [comment]
Weird that the tests seem flakey with setup-go v4.

**Comment by SuperQ** [comment]
Should I revert setup-go to v3.2.0 for now?

**Comment by dolmen** [comment]
Yes, let's stay on v3.2.0 for now.

**Comment by SuperQ** [comment]
All :green_circle:!

**Review by dolmen — APPROVED** [review]
LGTM

**Comment by SuperQ** [comment]
Thanks!

**Pull request merged**